saq/queue.py

```diff
@@ -394,12 +437,141 @@ async def enqueue(self, job_or_func, **kwargs):
         job.queued = now()
         job.status = Status.QUEUED
 
-        if not await self._enqueue_script(
-            keys=[self._incomplete, job.id, self._queued, job.abort_id],
-            args=[self.serialize(job), job.scheduled],
-            client=self.redis,
-        ):
-            return None
+        for cb in self._before_enqueues.values():
+            await cb(job)
+
+        async with self._enqueue_sem:
+            if not await self._enqueue_script(
+                keys=[self._incomplete, job.id, self._queued, job.abort_id],
+                args=[self.serialize(job), job.scheduled],
+                client=self.redis,
+            ):
+                return None
 
         logger.info("Enqueuing %s", job)
         return job
+
+    async def apply(self, job_or_func, timeout=None, **kwargs):
+        """
+        Enqueue a job and wait for its result.
+
+        If the job is successful, this returns its result.
+        If the job is unsuccessful, this raises a JobError.
+
+        Example::
+            try:
+                assert await queue.apply("add", a=1, b=2) == 3
+            except JobError:
+                print("job failed")
+
+        job_or_func: Same as Queue.enqueue
+        kwargs: Same as Queue.enqueue
+        """
+        results = await self.map(job_or_func, timeout=timeout, iter_kwargs=[kwargs])
+        return results[0]
+
+    async def map(
+        self, job_or_func, iter_kwargs, timeout=None, return_exceptions=False, **kwargs
+    ):
+        """
+        Enqueue multiple jobs and collect all of their results.
+
+        Example::
+            try:
+                assert await queue.map(
+                    "add",
+                    [
+                        {"a": 1, "b": 2},
+                        {"a": 3, "b": 4},
+                    ]
+                ) == [3, 7]
+            except JobError:
+                print("any of the jobs failed")
+
+        job_or_func: Same as Queue.enqueue
+        iter_kwargs: Enqueue a job for each item in this sequence. Each item is the same
+            as kwargs for Queue.enqueue.
+        timeout: Total seconds to wait for all jobs to complete. If None (default) or 0, wait forever.
+        return_exceptions: If False (default), an exception is immediately raised as soon as any jobs
+            fail. Other jobs won't be cancelled and will continue to run.
+            If True, exceptions are treated the same as successful results and aggregated in the result list.
+        kwargs: Default kwargs for all jobs. These will be overridden by those in iter_kwargs.
+        """
+        iter_kwargs = [
+            {"timeout": timeout, "key": kwargs.get("key") or uuid1(), **kwargs, **kw}
+            for kw in iter_kwargs
+        ]
+        job_ids = [Job.id_from_key(key["key"]) for key in iter_kwargs]
+        pending_job_ids = set(job_ids)
+
+        async def callback(job_id, status):
+            nonlocal pending_job_ids
```

Comment:
is this necessary? you aren't assigng it, so reference should be ok

Reply:
Ah, you're correct. Just leftover crud from when this was a counter. Good catch.